woocommerce/woocommerce
# Monorepo Utils: Add workflow profiler command

### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





This PR adds a CLI command to profile workflows.

<img width="1409" alt="image" src="https://github.com/woocommerce/woocommerce/assets/1922453/376bbeef-b4d3-4942-892d-f222a3c5445f">

#### Assumptions made

* Only successful workflow run are used to compile results
* All workflow runs lasting longer than 60 minutes are excluded.
* 90th percentile means the 90% slowest run



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. `pnpm install` to make sure the tool is correctly built.
2. Try running the tool for `Lint and tests for JS packages and woocommerce-admin/client`.
```
pnpm utils workflows profile 2023-06-01 2023-06-21 pr-lint-test-js.yml
```
2. Try again but pass `-s` to show job and step results too.
```
pnpm utils workflows profile 2023-06-01 2023-06-21 pr-lint-test-js.yml -s
```
3. Now run the tool for `Run tests against PR`, a workflow that has 6 jobs.
```
pnpm utils workflows profile 2023-06-20 2023-06-20 pr-build-and-e2e-tests.yml 
```
4. Now try running the tool with a fair amount of worflows. Here is `Lint and tests for JS packages and woocommerce-admin/client` for the month of May with 1003 runs.
```
pnpm utils workflows profile 2023-05-01 2023-05-31 pr-lint-test-js.yml -s
```
